test/e2e_yaml_test.go

```diff
@@ -0,0 +1,238 @@
+// +build e2e,yaml
+
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"io"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	knativetest "knative.dev/pkg/test"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+const (
+	DEFAULT_KO_DOCKER_REPO = `gcr.io\/christiewilson-catfactory`
+	DEFAULT_NAMESPACE      = `namespace: default`
+)
+
+// Generalized for tekton CRDs
+func GetCreatedTektonCrd(input []byte, kind string) (string, error) {
+	re := regexp.MustCompile(kind + `.tekton.dev\/(.+) created`)
+	submatch := re.FindSubmatch(input)
+	if submatch == nil {
+		return "", nil
+	}
+	return string(submatch[1]), nil
+}
+
+func ValidatePipelineRunDone(t *testing.T, c *clients, pipelineRunName string) {
+	err := WaitForPipelineRunState(c, pipelineRunName, pipelineRunTimeout, Succeed(pipelineRunName), pipelineRunName)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func ValidateTaskRunDone(t *testing.T, c *clients, taskRunName string) {
+	// Per test basis
+	err := WaitForTaskRunState(c, taskRunName, Succeed(taskRunName), taskRunName)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+// Substitute docker repos/bucket paths from environment to allow tests on
+// local clusters. Also update namespaces so that examples with ServiceAccounts
+// work under test
+func SubEnv(input []byte, namespace string) ([]byte, error) {
+	val, ok := os.LookupEnv("KO_DOCKER_REPO")
+	var output []byte
+	if ok {
+		re := regexp.MustCompile(DEFAULT_KO_DOCKER_REPO)
+		output = re.ReplaceAll(input, []byte(val))
+	} else {
+		return input, nil
+	}
+
+	re := regexp.MustCompile(DEFAULT_NAMESPACE)
+	output = re.ReplaceAll(output, []byte(strings.ReplaceAll(DEFAULT_NAMESPACE, "default", namespace)))
```

Comment:
i think it might be a bit simpler to provide the namespace argument to ko instead: you can use `--namespace` as an arg to ko and then afaik you wont need to do this replacement

Reply:
I'd added this to fix the use of ServiceAccount in  [`v1alpha1/pipelineruns/pipelinerun.yaml`](https://github.com/tektoncd/pipeline/blob/master/examples/v1alpha1/pipelineruns/pipelinerun.yaml) and [`v1beta1/pipelineruns/pipelinerun.yaml`](https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/pipelinerun.yaml), which specifies the `default` namespace in the examples.

Is it possible to override the ServiceAccount namespace with the `ko` arg? I thought much like `kubectl`, the hardcoded namespace in the example would still apply. However, my understanding is a bit limited!